File: typescript/packages/vox-core/src/metadata.ts
// Client-side metadata: a self-describing `Value` map (`r[rpc.metadata]`).
//
// On the wire metadata is a phon `Value` map; flags are well-known keys whose
// value is a list of the key names they apply to (`r[rpc.metadata.flags]`).

import type { Value } from "@bearcove/phon-schema";
import {
  type Metadata,
  MetadataKeys,
  metadataAddFlag,
  metadataIsSensitive,
  metadataIsNoPropagate,
} from "@bearcove/vox-wire";

/** A metadata value: string, u64 (bigint), or raw bytes. */
export type ClientMetadataValue = string | bigint | Uint8Array;

/**
 * Client-side metadata builder.
 *
 * `set()` for normal metadata; `setSensitive()` marks a key for redaction in logs
 * and traces (recorded under the `vox:sensitive` well-known key).
 */
export class ClientMetadata {
  private readonly map: Metadata = new Map();

  set(key: string, value: ClientMetadataValue): this {
    this.map.set(key, value as Value);
    return this;
  }

  /** r[impl rpc.metadata.flags.sensitive] */
  setSensitive(key: string, value: ClientMetadataValue): this {
    this.map.set(key, value as Value);
    metadataAddFlag(this.map, MetadataKeys.SENSITIVE, key);
    return this;
  }

  /** r[impl rpc.metadata.flags.no-propagate] */
  setNoPropagate(key: string, value: ClientMetadataValue): this {
    this.map.set(key, value as Value);
    metadataAddFlag(this.map, MetadataKeys.NO_PROPAGATE, key);
    return this;
  }

  get(key: string): Value | undefined {
    return this.map.get(key);
  }

  has(key: string): boolean {
    return this.map.has(key);
  }

  delete(key: string): boolean {
    return this.map.delete(key);
  }

  get size(): number {
    return this.map.size;
  }

  keys(): IterableIterator<string> {
    return this.map.keys();
  }

  entries(): IterableIterator<[string, Value]> {
    return this.map.entries();
  }

  isSensitive(key: string): boolean {
    return metadataIsSensitive(this.map, key);
  }

  isNoPropagate(key: string): boolean {
    return metadataIsNoPropagate(this.map, key);
  }

  /** The wire `Value` map (flags already folded into well-known keys). */
  toWire(): Metadata {
    return this.map;
  }

  clone(): ClientMetadata {
    const copy = new ClientMetadata();
    for (const [k, v] of this.map) copy.map.set(k, v);
    return copy;
  }

  static fromWire(metadata: Metadata): ClientMetadata {
    const m = new ClientMetadata();
    for (const [k, v] of metadata) m.map.set(k, v);
    return m;
  }
}

/** Convert a `ClientMetadata` to the wire `Value` map. */
export function clientMetadataToWire(metadata: ClientMetadata): Metadata {
  return metadata.toWire();
}

File: swift/vox-runtime/Sources/VoxRuntime/Metadata.swift
import PhonSchema

// Metadata: a self-describing key→value map carried on the wire as a phon dynamic
// `Value` (`r[rpc.metadata]`) — an object of string keys to string / bytes / u64
// values, or null when empty. Mirrors `rust/vox-types/src/metadata.rs`.
// r[impl schema.interaction.metadata]
//
// Per-key behavior flags (sensitive, no-propagate) are recorded under the
// well-known keys below, each holding an array of the metadata key-names the flag
// applies to — they are NOT a per-entry wire field.

/// Metadata is a self-describing [`Value`].
public typealias Metadata = Value

/// Well-known metadata keys carrying per-key behavior flags
/// (`r[rpc.metadata.flags]`).
public enum MetadataKeys {
    /// Keys whose values MUST NOT be logged/traced (`r[rpc.metadata.flags.sensitive]`).
    public static let sensitive = "vox:sensitive"
    /// Keys whose values MUST NOT be forwarded downstream
    /// (`r[rpc.metadata.flags.no-propagate]`).
    public static let noPropagate = "vox:no-propagate"
}

/// Empty metadata reads as null (an absent metadata field).
public func emptyMetadata() -> Metadata { .null }

/// Coerce a decoded value to metadata: anything that is not an object reads as empty.
public func coerceMetadata(_ value: Value) -> Metadata {
    if case .object = value { return value }
    return .null
}

public extension Value {
    /// The string value at `key`, if present and a string.
    func metaStr(_ key: String) -> String? { get(key)?.asString }

    /// The `u64` value at `key`, if present and a non-negative integer that fits.
    func metaU64(_ key: String) -> UInt64? { get(key)?.asNumber?.toU64 }

    /// The byte-run value at `key`, if present and bytes.
    func metaBytes(_ key: String) -> [UInt8]? { get(key)?.asBytes }

    /// Whether there are no metadata entries.
    var metaIsEmpty: Bool { (asObject?.isEmpty) ?? true }

    /// The number of entries (0 when null).
    var metaLen: Int { asObject?.count ?? 0 }

    /// Whether a key is present.
    func metaHas(_ key: String) -> Bool { get(key) != nil }

    /// The (non-well-known) `(key, value)` entries.
    func metaEntries() -> [(String, Value)] {
        guard case .object(let entries) = self else { return [] }
        return entries
            .filter { $0.key != MetadataKeys.sensitive && $0.key != MetadataKeys.noPropagate }
            .map { ($0.key, $0.value) }
    }

    /// Insert (or replace) `key`→`value`, creating the object if needed. The shared
    /// construction primitive (mirrors `meta_set`).
    mutating func metaSet(_ key: String, _ value: Value) {
        var entries: [Entry]
        if case .object(let e) = self { entries = e } else { entries = [] }
        if let i = entries.firstIndex(where: { $0.key == key }) {
            entries[i].value = value
        } else {
            entries.append(Entry(key: key, value: value))
        }
        self = .object(entries)
    }

    /// Remove `key` if present.
    mutating func metaRemove(_ key: String) {
        guard case .object(var entries) = self else { return }
        entries.removeAll { $0.key == key }
        self = entries.isEmpty ? .null : .object(entries)
    }

    /// Functional `metaSet` returning a new metadata value.
    func metaSetting(_ key: String, _ value: Value) -> Metadata {
        var copy = self
        copy.metaSet(key, value)
        return copy
    }
}

/// A `u64` metadata value (canonicalized like the Rust/TS number model).
public func metaU64Value(_ v: UInt64) -> Value { .number(.canonical(unsigned: UInt128(v))) }
